File: app/src/main/java/com/kispoko/tome/sheet/widget/ExpanderWidget.java

package com.kispoko.tome.sheet.widget;


import android.content.Context;
import android.view.Gravity;
import android.view.View;
import android.widget.LinearLayout;
import android.widget.RelativeLayout;

import com.kispoko.tome.R;
import com.kispoko.tome.sheet.Alignment;
import com.kispoko.tome.sheet.BackgroundColor;
import com.kispoko.tome.sheet.Corners;
import com.kispoko.tome.sheet.Spacing;
import com.kispoko.tome.sheet.group.Group;
import com.kispoko.tome.sheet.group.GroupParent;
import com.kispoko.tome.sheet.widget.expander.ExpanderWidgetFormat;
import com.kispoko.tome.sheet.widget.util.WidgetData;
import com.kispoko.tome.lib.ui.ImageViewBuilder;
import com.kispoko.tome.lib.ui.LayoutType;
import com.kispoko.tome.lib.ui.LinearLayoutBuilder;
import com.kispoko.tome.lib.ui.RelativeLayoutBuilder;
import com.kispoko.tome.lib.ui.TextViewBuilder;
import com.kispoko.tome.lib.functor.CollectionFunctor;
import com.kispoko.tome.lib.functor.ModelFunctor;
import com.kispoko.tome.lib.functor.PrimitiveFunctor;
import com.kispoko.tome.lib.yaml.YamlBuilder;
import com.kispoko.tome.lib.yaml.YamlParseException;
import com.kispoko.tome.lib.yaml.YamlParser;

import java.io.Serializable;
import java.util.List;
import java.util.UUID;



/**
 * Expander Widget
 */
public class ExpanderWidget extends Widget
                            implements Serializable
{

    // PROPERTIES
    // -----------------------------------------------------------------------------------------

    // > Model
    // -----------------------------------------------------------------------------------------

    private UUID                                id;


    // > Functors
    // -----------------------------------------------------------------------------------------

    private PrimitiveFunctor<String>            name;
    private CollectionFunctor<Group>            groups;
    private ModelFunctor<ExpanderWidgetFormat>  format;
    private ModelFunctor<WidgetData>            widgetData;


    // CONSTRUCTORS
    // -----------------------------------------------------------------------------------------


    public ExpanderWidget()
    {
        this.id         = null;

        this.name       = new PrimitiveFunctor<>(null, String.class);
        this.groups     = CollectionFunctor.empty(Group.class);
        this.format     = ModelFunctor.empty(ExpanderWidgetFormat.class);
        this.widgetData = ModelFunctor.empty(WidgetData.class);
    }


    public ExpanderWidget(UUID id,
                          String name,
                          List<Group> groups,
                          ExpanderWidgetFormat format,
                          WidgetData widgetData)
    {
        this.id         = id;

        this.name       = new PrimitiveFunctor<>(name, String.class);
        this.groups     = CollectionFunctor.full(groups, Group.class);
        this.format     = ModelFunctor.full(format, ExpanderWidgetFormat.class);
        this.widgetData = ModelFunctor.full(widgetData, WidgetData.class);
    }


    /**
     * Create a new Expander Widget from its yaml representation.
     * @param yaml The yaml parser.
     * @return The parsed Exapnder Widget.
     * @throws YamlParseException
     */
    public static ExpanderWidget fromYaml(YamlParser yaml)
                  throws YamlParseException
    {
        UUID                 id         = UUID.randomUUID();

        String               name       = yaml.atKey("name").getString();

        List<Group>          groups     = yaml.atMaybeKey("groups").forEach(
                                                            new YamlParser.ForEach<Group>() {
            @Override
            public Group forEach(YamlParser yaml, int index) throws YamlParseException {
                return Group.fromYaml(yaml, index);
            }
        }, true);

        ExpanderWidgetFormat format     = ExpanderWidgetFormat.fromYaml(yaml.atMaybeKey("format"));

        WidgetData           widgetData = WidgetData.fromYaml(yaml.atMaybeKey("data"));

        return new ExpanderWidget(id, name, groups, format, widgetData);
    }


    // API
    // ------------------------------------------------------------------------------------------

    // > Model
    // ------------------------------------------------------------------------------------------

    // ** Id
    // ------------------------------------------------------------------------------------------

    /**
     * Get the model identifier.
     * @return The model UUID.
     */
    public UUID getId()
    {
        return this.id;
    }


    /**
     * Set the model identifier.
     * @param id The new model UUID.
     */
    public void setId(UUID id)
    {
        this.id = id;
    }


    // ** On Load
    // ------------------------------------------------------------------------------------------

    /**
     * This method is called when the RulesEngine is completely loaded for the first time.
     */
    public void onLoad()
    {
        this.initializeExpanderWidget();
    }


    // > Yaml
    // ------------------------------------------------------------------------------------------

    /**
     * The Expander Widget's yaml representation.
     * @return The yaml builder.
     */
    public YamlBuilder toYaml()
    {
        return YamlBuilder.map()
                .putString("name", this.name())
                .putList("groups", this.groups())
                .putYaml("format", this.format())
                .putYaml("data", this.data());
    }


    // > Widget
    // ------------------------------------------------------------------------------------------

    @Override
    public void initialize(GroupParent groupParent, Context context) { }


    @Override
    public WidgetData data()
    {
        return this.widgetData.getValue();
    }


    /**
     * The text widget's tile view.
     * @return The tile view.
     */
    @Override
    public View view(boolean rowHasLabel, Context context)
    {
        return this.widgetView(rowHasLabel, context);
    }


    // > State
    // ------------------------------------------------------------------------------------------

    /**
     * The expander name.
     * @return The name.
     */
    public String name()
    {
        return this.name.getValue();
    }


    /**
     * The groups that are expanded/hidden.
     * @return The groups.
     */
    public List<Group> groups()
    {
        return this.groups.getValue();
    }


    /**
     * The expander widget formatting options.
     * @return The format.
     */
    public ExpanderWidgetFormat format()
    {
        return this.format.getValue();
    }


    // INTERNAL
    // -----------------------------------------------------------------------------------------

    // > Initialize
    // -----------------------------------------------------------------------------------------

    private void initializeExpanderWidget()
    {
        // [1] Apply default formats
        // -------------------------------------------------------------------------------------

        if (this.data().format().alignmentIsDefault())
            this.data().format().setAlignment(Alignment.CENTER);

        if (this.data().format().backgroundIsDefault())
            this.data().format().setBackground(BackgroundColor.NONE);

        if (this.data().format().cornersIsDefault())
            this.data().format().setCorners(Corners.NONE);
    }


    // > Views
    // -----------------------------------------------------------------------------------------

    private LinearLayout widgetView(boolean rowHasLabel, Context context)
    {
        LinearLayout layout = this.layout(rowHasLabel, context);

        layout.addView(mainView(context));

        return layout;
    }


    private LinearLayout mainView(Context context)
    {
        LinearLayout layout = mainViewLayout(context);

        // > Header View
        // -----------------------------------------------------------
        layout.addView(headerView(context));

        return layout;
    }


    private LinearLayout mainViewLayout(Context context)
    {
        LinearLayoutBuilder layout = new LinearLayoutBuilder();

        layout.orientation          = LinearLayout.VERTICAL;
        layout.width                = LinearLayout.LayoutParams.MATCH_PARENT;
        layout.height               = LinearLayout.LayoutParams.WRAP_CONTENT;

        layout.backgroundColor      = this.data().format().background().colorId();
        layout.backgroundResource   = this.data().format().corners().resourceId();


        Spacing margins = this.data().format().margins();

        if (this.data().format().elevation() != null)
        {
            layout.elevation        = this.data().format().elevation().floatValue();

            if (margins.bottom() < 3)
                margins.setBottom(3);

            if (margins.left() < 3)
                margins.setLeft(3);

            if (margins.right() < 3)
                margins.setRight(3);
        }

        layout.marginSpacing        = margins;

        return layout.linearLayout(context);
    }


    private LinearLayout headerView(Context context)
    {
        LinearLayout layout = headerViewLayout(context);

        this.onExpanderClick(layout);

        // > Name View
        layout.addView(nameView(context));


        return layout;
    }


    private LinearLayout headerViewLayout(Context context)
    {
        LinearLayoutBuilder layout = new LinearLayoutBuilder();

        layout.orientation      = LinearLayout.VERTICAL;
        layout.width            = LinearLayout.LayoutParams.MATCH_PARENT;
        layout.height           = LinearLayout.LayoutParams.WRAP_CONTENT;

        layout.paddingSpacing   = this.format().headerPadding();

        return layout.linearLayout(context);
    }

    private RelativeLayout nameView(Context context)
    {
        // [1] Declarations
        // -------------------------------------------------------------------------------------

        RelativeLayoutBuilder layout  = new RelativeLayoutBuilder();

        TextViewBuilder       name    = new TextViewBuilder();
        ImageViewBuilder      icon    = new ImageViewBuilder();

        // [2] Layout
        // -------------------------------------------------------------------------------------

        layout.width                = LinearLayout.LayoutParams.MATCH_PARENT;
        layout.height               = LinearLayout.LayoutParams.WRAP_CONTENT;

        layout.gravity              = Gravity.CENTER_VERTICAL;

        layout.addRule(RelativeLayout.CENTER_VERTICAL);

        layout.child(name)
              .child(icon);

        // [3 A] Name
        // -------------------------------------------------------------------------------------

        name.layoutType         = LayoutType.RELATIVE;
        name.width              = RelativeLayout.LayoutParams.WRAP_CONTENT;
        name.height             = RelativeLayout.LayoutParams.WRAP_CONTENT;

        name.text               = this.name();

        name.addRule(RelativeLayout.ALIGN_PARENT_LEFT);
        name.addRule(RelativeLayout.CENTER_VERTICAL);

        this.format().nameStyleClosed().styleTextViewBuilder(name, context);

        // [3 B] Icon
        // -------------------------------------------------------------------------------------

        icon.layoutType         = LayoutType.RELATIVE;
        icon.width              = RelativeLayout.LayoutParams.WRAP_CONTENT;
        icon.height             = RelativeLayout.LayoutParams.WRAP_CONTENT;

        icon.image              = R.drawable.ic_expander_more;

        //icon.color              = this.format().nameStyleClosed().color().resourceId();
        icon.color              = R.color.dark_blue_hl_8;

        icon.addRule(RelativeLayout.ALIGN_PARENT_RIGHT);
        icon.addRule(RelativeLayout.CENTER_VERTICAL);


        return layout.relativeLayout(context);
    }


    private void onExpanderClick(LinearLayout layout)
    {
        layout.setOnClickListener(new View.OnClickListener()
        {
            @Override
            public void onClick(View view)
            {

            }
        });
    }


}

File: app/src/main/java/com/taletable/android/model/sheet/widget/expander/ExpanderWidget.kt

package com.taletable.android.model.sheet.widget.expander


import android.content.Context
import android.view.View
import android.view.ViewGroup
import com.taletable.android.lib.Factory
import com.taletable.android.lib.orm.sql.SQLSerializable
import com.taletable.android.lib.orm.sql.SQLText
import com.taletable.android.lib.orm.sql.SQLValue
import com.taletable.android.model.sheet.group.GroupContext
import com.taletable.android.model.sheet.group.RowLayoutType
import com.taletable.android.model.sheet.widget.ExpanderWidget
import com.taletable.android.model.sheet.widget.NumberWidget
import com.taletable.android.model.sheet.widget.Widget
import com.taletable.android.model.sheet.widget.WidgetView
import com.taletable.android.model.sheet.widget.number.NumberWidgetFormatCustom
import com.taletable.android.model.sheet.widget.number.NumberWidgetFormatOfficial
import com.taletable.android.model.sheet.widget.number.numberWidgetCustomView
import com.taletable.android.model.sheet.widget.number.numberWidgetOfficialView
import com.taletable.android.rts.entity.EntityId
import com.taletable.android.rts.entity.EntityTypeSheet
import com.taletable.android.rts.entity.entityType
import effect.*
import lulo.document.*
import lulo.value.UnexpectedType
import lulo.value.UnexpectedValue
import lulo.value.ValueError
import lulo.value.ValueParser
import maybe.Maybe
import maybe.Nothing
import java.io.Serializable



// -----------------------------------------------------------------------------------------
// | VIEW
// -----------------------------------------------------------------------------------------

/**
 * Expander Widget view
 */
fun expanderWidgetViewGroup(
        expanderWidget : ExpanderWidget,
        rowLayoutType : RowLayoutType,
        entityId : EntityId,
        context : Context
) : ViewGroup
{
    val layout = WidgetView.layout(expanderWidget.widgetFormat(), entityId, context, rowLayoutType)

//    configureNumberWidgetViewClick(layout, numberWidget, entityId, context)

    return layout
}

///**
// * Number Widget view
// */
//private fun configureNumberWidgetViewClick(
//        viewGroup : ViewGroup,
//        widget : Widget,
//        entityId : EntityId,
//        context : Context
//)
//{
//    entityType(entityId).apDo { entityType ->
//        when (entityType)
//        {
//            is EntityTypeSheet ->
//            {
//                // NORMAL CLICK
//                viewGroup.setOnClickListener {
//                    widget.primaryAction(entityId, context)
//                }
//                // LONG CLICK
//                viewGroup.setOnLongClickListener {
//                    widget.secondaryAction(entityId, context)
//                    true
//                }
//            }
//        }
//
//    }
//}


fun expanderWidgetView(
        expanderWidget : ExpanderWidget,
        entityId : EntityId,
        context : Context,
        groupContext : Maybe<GroupContext> = Nothing()) : View
{
    val format = expanderWidget.format

    return when (format) {
        is ExpanderWidgetFormatCustom ->
            expanderWidgetCustomView(expanderWidget, format, entityId, context, groupContext)
        is ExpanderWidgetFormatOfficial ->
            expanderWidgetOfficialView(format, expanderWidget, entityId, context, groupContext)
    }
}



/**
 * Expander Widget View Type
 */
sealed class ExpanderWidgetViewType : ToDocument, SQLSerializable, Serializable
{

    object Plain : ExpanderWidgetViewType()
    {
        // SQL SERIALIZABLE
        // -------------------------------------------------------------------------------------

        override fun asSQLValue() : SQLValue = SQLText({ "plain" })

        // TO DOCUMENT
        // -------------------------------------------------------------------------------------

        override fun toDocument() = DocText("plain")

    }


    object IconLeft : ExpanderWidgetViewType()
    {
        // SQL SERIALIZABLE
        // -------------------------------------------------------------------------------------

        override fun asSQLValue() : SQLValue = SQLText({ "icon_left" })

        // TO DOCUMENT
        // -------------------------------------------------------------------------------------

        override fun toDocument() = DocText("icon_left")

    }


    object IconRight : ExpanderWidgetViewType()
    {
        // SQL SERIALIZABLE
        // -------------------------------------------------------------------------------------

        override fun asSQLValue() : SQLValue = SQLText({ "icon_right" })

        // TO DOCUMENT
        // -------------------------------------------------------------------------------------

        override fun toDocument() = DocText("icon_right")

    }


    object IconRightLink : ExpanderWidgetViewType()
    {
        // SQL SERIALIZABLE
        // -------------------------------------------------------------------------------------

        override fun asSQLValue() : SQLValue = SQLText({ "icon_right_link" })

        // TO DOCUMENT
        // -------------------------------------------------------------------------------------

        override fun toDocument() = DocText("icon_right_link")

    }


    object ButtonBottom : ExpanderWidgetViewType()
    {
        // SQL SERIALIZABLE
        // -------------------------------------------------------------------------------------

        override fun asSQLValue() : SQLValue = SQLText({ "button_bottom" })

        // TO DOCUMENT
        // -------------------------------------------------------------------------------------

        override fun toDocument() = DocText("button_bottom")

    }


    object CircleAvatarLeft : ExpanderWidgetViewType()
    {
        // SQL SERIALIZABLE
        // -------------------------------------------------------------------------------------

        override fun asSQLValue() : SQLValue = SQLText({ "circle_avatar_left" })

        // TO DOCUMENT
        // -------------------------------------------------------------------------------------

        override fun toDocument() = DocText("circle_avatar_left")

    }


    object CheckboxLeft : ExpanderWidgetViewType()
    {
        // SQL SERIALIZABLE
        // -------------------------------------------------------------------------------------

        override fun asSQLValue() : SQLValue = SQLText({ "checkbox_left" })

        // TO DOCUMENT
        // -------------------------------------------------------------------------------------

        override fun toDocument() = DocText( "checkbox_left" )

    }


    // -----------------------------------------------------------------------------------------
    // CONSTRUCTORS
    // -----------------------------------------------------------------------------------------

    companion object
    {
        fun fromDocument(doc : SchemaDoc) : ValueParser<ExpanderWidgetViewType> = when (doc)
        {
            is DocText -> when (doc.text)
            {
                "icon_left"          -> effValue<ValueError, ExpanderWidgetViewType>(
                        IconLeft)
                "icon_right"         -> effValue<ValueError, ExpanderWidgetViewType>(
                        IconRight)
                "icon_right_link"    -> effValue<ValueError, ExpanderWidgetViewType>(
                        IconRightLink)
                "plain"              -> effValue<ValueError, ExpanderWidgetViewType>(
                        Plain)
                "button_bottom"      -> effValue<ValueError, ExpanderWidgetViewType>(
                        ButtonBottom)
                "circle_avatar_left" -> effValue<ValueError, ExpanderWidgetViewType>(
                        CircleAvatarLeft)
                "checkbox_left"      -> effValue<ValueError, ExpanderWidgetViewType>(
                        CheckboxLeft)
                else            -> effError<ValueError, ExpanderWidgetViewType>(
                                            UnexpectedValue("ExpanderWidgetViewType", doc.text, doc.path))
            }
            else       -> effError(lulo.value.UnexpectedType(DocType.TEXT, docType(doc), doc.path))
        }
    }

}


/**
 * Expander Widget Label
 */
data class ExpanderWidgetLabel(val value : String) : ToDocument, SQLSerializable, Serializable
{

    // -----------------------------------------------------------------------------------------
    // CONSTRUCTORS
    // -----------------------------------------------------------------------------------------

    companion object : Factory<ExpanderWidgetLabel>
    {
        override fun fromDocument(doc: SchemaDoc): ValueParser<ExpanderWidgetLabel> = when (doc)
        {
            is DocText -> effValue(ExpanderWidgetLabel(doc.text))
            else       -> effError(UnexpectedType(DocType.TEXT, docType(doc), doc.path))
        }
    }


    // -----------------------------------------------------------------------------------------
    // TO DOCUMENT
    // -----------------------------------------------------------------------------------------

    override fun toDocument() = DocText(this.value)


    // -----------------------------------------------------------------------------------------
    // SQL SERIALIZABLE
    // -----------------------------------------------------------------------------------------

    override fun asSQLValue() : SQLValue = SQLText({this.value})

}


//class ExpanderWidgetUI(val expanderWidget : ExpanderWidget,
//                       val entityId : EntityId,
//                       val context : Context,
//                       val groupContext : Maybe<GroupContext> = Nothing())
//{
//
//    // -----------------------------------------------------------------------------------------
//    // PROPERTIES
//    // -----------------------------------------------------------------------------------------
//
//    val activity = context as AppCompatActivity
//
//    var isOpen : Boolean = false
//
//    var groupsLayout : LinearLayout? = null
//
//    var iconView : ImageView? = null
//    var showMoreView : TextView? = null
//
//    var checkboxLayout : LinearLayout? = null
//
//
//    // | Checkbox
//    // -----------------------------------------------------------------------------------------
//
//
//    fun select(entityId : EntityId)
//    {
//    }
//
//
//    // | Update
//    // -----------------------------------------------------------------------------------------
//
//    fun updateCheckboxView(layout : LinearLayout)
//    {
//        layout.removeAllViews()
//        layout.addView(this.checkboxView())
//    }
//
//
//    // -----------------------------------------------------------------------------------------
//    // VIEWS
//    // -----------------------------------------------------------------------------------------
//
//    fun view() : View
//    {
//        val layout = WidgetView.layout(expanderWidget.widgetFormat(), entityId, context)
//
//        val layoutId = Util.generateViewId()
//        layout.id = layoutId
//        expanderWidget.layoutId = layoutId
//
//        val contentLayout = layout.findViewById<LinearLayout>(R.id.widget_content_layout)
//        contentLayout.orientation       = LinearLayout.VERTICAL
//
//        // Header
//        val headerView = this.headerView()
//        headerView.setOnClickListener { onClick(contentLayout) }
//        contentLayout.addView(headerView)
//
//        Log.d("***EXPANDER WIDGET", "context is $groupContext")
//
//        return layout
//    }
//

//
//

//
//    // -----------------------------------------------------------------------------------------
//    // HEADER
//    // -----------------------------------------------------------------------------------------
//
//    private fun headerView() : View = when (expanderWidget.format().viewType())
//    {
//        is ExpanderWidgetViewType.Plain -> plainHeaderView()
//        is ExpanderWidgetViewType.IconLeft -> iconLeftHeaderView()
//        is ExpanderWidgetViewType.IconRight -> iconRightHeaderView()
//        is ExpanderWidgetViewType.IconRightLink -> iconRightLinkHeaderView()
//        is ExpanderWidgetViewType.ButtonBottom -> this.buttonBottomHeaderView()
//        is ExpanderWidgetViewType.CircleAvatarLeft -> this.circleAvatarLeftView()
//        is ExpanderWidgetViewType.CheckboxLeft -> this.checkboxLeftHeaderView()
//    }
//
//
//    // HEADER > Common Views
//    // -----------------------------------------------------------------------------------------
//
//    private fun headerLabelView() : TextView
//    {
//        val title           = TextViewBuilder()
//
//        val format          = if (this.isOpen)
//                                  expanderWidget.format().headerLabelOpenFormat()
//                              else
//                                  expanderWidget.format().headerLabelClosedFormat()
//
//        title.width         = LinearLayout.LayoutParams.WRAP_CONTENT
//        title.height        = LinearLayout.LayoutParams.WRAP_CONTENT
//
//        title.text          = expanderWidget.labelValue(entityId, groupContext)
//
//        format.styleTextViewBuilder(title, entityId, context)
//
//        title.color           = colorOrBlack(format.colorTheme(), entityId)
//
//        title.marginSpacing = format.elementFormat().margins()
//
//        when (expanderWidget.format().viewType()) {
//            is ExpanderWidgetViewType.CheckboxLeft -> {
//                title.padding.topDp    = format.elementFormat().padding().topDp()
//                title.padding.bottomDp = format.elementFormat().padding().bottomDp()
//                title.padding.rightDp   = format.elementFormat().padding().rightDp()
//            }
//            else -> {
//                title.paddingSpacing = format.elementFormat().padding()
//            }
//        }
//
//        return title.textView(context)
//    }
//
//
//    private fun headerViewLayout() : LinearLayout
//    {
//        val layout              = LinearLayoutBuilder()
//
//        val format = if (this.isOpen)
//            expanderWidget.format().headerOpenFormat()
//        else
//            expanderWidget.format().headerClosedFormat()
//
//        layout.width            = LinearLayout.LayoutParams.MATCH_PARENT
//        layout.height           = LinearLayout.LayoutParams.WRAP_CONTENT
//
//        layout.orientation      = LinearLayout.HORIZONTAL
//
//        layout.gravity          = Gravity.CENTER_VERTICAL
//
//        layout.marginSpacing    = format.elementFormat().margins()
//
//        // Padding
//        when (expanderWidget.format().viewType()) {
//            is ExpanderWidgetViewType.CheckboxLeft -> {
//                //layout.padding.leftDp   = format.elementFormat().padding().leftDp()
//                layout.padding.rightDp   = format.elementFormat().padding().rightDp()
//            }
//            else -> {
//                layout.paddingSpacing   = format.elementFormat().padding()
//            }
//        }
//
//        layout.corners          = format.elementFormat().corners()
//        // layout.corners          = Corners(2.0, 2.0, 2.0, 2.0)
//
//        layout.backgroundColor  = colorOrBlack(format.elementFormat().backgroundColorTheme(),
//                                               entityId)
//
//        expanderWidget.bookReference.doMaybe { bookRef ->
//            layout.onLongClick = View.OnLongClickListener {
//                val intent = Intent(activity, BookActivity::class.java)
//                intent.putExtra("book_reference", bookRef)
//                activity.startActivity(intent)
//                true
//            }
//        }
//
//        return layout.linearLayout(context)
//    }
//
//
//    // HEADER > Plain
//    // -----------------------------------------------------------------------------------------
//
//    private fun plainHeaderView() : LinearLayout
//    {
//        val layout = this.headerViewLayout()
//
//        layout.addView(this.headerLabelView())
//
//        return layout
//    }
//
//
//    // HEADER > Left Icon
//    // -----------------------------------------------------------------------------------------
//
//    private fun iconLeftHeaderView() : LinearLayout
//    {
//        val layout          = this.headerViewLayout()
//
//        // Icon
//        val iconLayout = this.headerIconLayoutView()
//        val iconView = this.headerIconView()
//        this.iconView = iconView
//        iconLayout.addView(iconView)
//        layout.addView(iconLayout)
//
//        // Label
//        layout.addView(this.headerLabelView())
//
//        return layout
//    }
//
//
//
//    private fun headerIconLayoutView() : LinearLayout
//    {
//        val layout = LinearLayoutBuilder()
//
//        val format = if (this.isOpen)
//            expanderWidget.format().headerOpenIcon().elementFormat()
//        else
//            expanderWidget.format().headerClosedIcon().elementFormat()
//
//        layout.width                = LinearLayout.LayoutParams.WRAP_CONTENT
//        layout.height               = LinearLayout.LayoutParams.WRAP_CONTENT
//
//        layout.backgroundColor      = colorOrBlack(format.backgroundColorTheme(),
//                                                   entityId)
//
//        layout.corners              = format.corners()
//
//        layout.paddingSpacing       = format.padding()
//        layout.marginSpacing        = format.margins()
//
//        return layout.linearLayout(context)
//    }
//
//
//    private fun headerIconView(iconId : Int? = null) : ImageView
//    {
//        val icon = ImageViewBuilder()
//
//        val format = if (this.isOpen)
//            expanderWidget.format().headerOpenIcon().iconFormat()
//        else
//            expanderWidget.format().headerClosedIcon().iconFormat()
//
//        icon.widthDp        = format.size().width
//        icon.heightDp       = format.size().height
//
//        if (this.isOpen) {
//            if (iconId != null) {
//                icon.image = iconId
//            } else {
//                icon.image      = expanderWidget.format().headerOpenIcon().iconType().drawableResId()
//            }
//        }
//        else {
//            if (iconId != null) {
//                icon.image  = iconId
//            } else {
//                icon.image      = expanderWidget.format().headerClosedIcon().iconType().drawableResId()
//            }
//        }
//
//        icon.color          = colorOrBlack(format.colorTheme(), entityId)
//
//        return icon.imageView(context)
//    }
//
//
//    // HEADER > Right Icon
//    // -----------------------------------------------------------------------------------------
//
//    private fun iconRightHeaderView() : LinearLayout
//    {
//        val layout          = this.headerViewLayout()
//
//        // Label
//        val labelView = this.headerLabelView()
//        val labelLayoutParams = labelView.layoutParams as LinearLayout.LayoutParams
//        labelLayoutParams.width = 0
//        labelLayoutParams.height = LinearLayout.LayoutParams.WRAP_CONTENT
//        labelLayoutParams.weight  = 1f
//        labelView.layoutParams = labelLayoutParams
//
//        layout.addView(labelView)
//
//        // Icon
//        val iconLayout = this.headerIconLayoutView()
//
//        val iconView = this.headerIconView()
//        this.iconView = iconView
//        iconLayout.addView(iconView)
//        layout.addView(iconLayout)
//
//        return layout
//    }
//
//
//    // HEADER > Right Icon Link
//    // -----------------------------------------------------------------------------------------
//
//    private fun iconRightLinkHeaderView() : LinearLayout
//    {
//        val layout          = this.headerViewLayout()
//
//        // Label
//        val labelView = this.headerLabelView()
//        val labelLayoutParams = labelView.layoutParams as LinearLayout.LayoutParams
//        labelLayoutParams.width = 0
//        labelLayoutParams.height = LinearLayout.LayoutParams.WRAP_CONTENT
//        labelLayoutParams.weight  = 1f
//        labelView.layoutParams = labelLayoutParams
//
//        layout.addView(labelView)
//
//        // Icon
//        val iconLayout = this.headerIconLayoutView()
//
//        val iconView = this.headerIconView(R.drawable.icon_open_in_window)
//        this.iconView = iconView
//        iconLayout.addView(iconView)
//        layout.addView(iconLayout)
//
//        return layout
//    }
//
//
//
//    // HEADER > Button Bottom
//    // -----------------------------------------------------------------------------------------
//
//    private fun buttonBottomHeaderView() : LinearLayout
//    {
//        val layout          = this.buttonBottomHeaderViewLayout()
//
//        // Label
//        layout.addView(this.headerLabelView())
//
//        // Icon
////        val showMoreView = this.showMoreButtonView()
////        this.showMoreView = showMoreView
////        layout.addView(showMoreView)
//
//        return layout
//    }
//
//
//    private fun buttonBottomHeaderViewLayout() : LinearLayout
//    {
//        val layout              = LinearLayoutBuilder()
//
//        val format = if (this.isOpen)
//            expanderWidget.format().headerOpenFormat()
//        else
//            expanderWidget.format().headerClosedFormat()
//
//        layout.width            = LinearLayout.LayoutParams.MATCH_PARENT
//        layout.height           = LinearLayout.LayoutParams.WRAP_CONTENT
//
//        layout.orientation      = LinearLayout.VERTICAL
//
//        layout.paddingSpacing   = format.elementFormat().padding()
//        layout.marginSpacing    = format.elementFormat().margins()
//
//        layout.corners          = Corners(2.0, 2.0, 2.0, 2.0)
//
//        layout.backgroundColor  = colorOrBlack(format.elementFormat().backgroundColorTheme(),
//                                               entityId)
//
//        return layout.linearLayout(context)
//    }
//
//
//    private fun circleAvatarLeftView() : LinearLayout
//    {
//        val headerFormat = if (this.isOpen)
//                               expanderWidget.format().headerOpenFormat()
//                           else
//                               expanderWidget.format().headerClosedFormat()
//
//        val layout = this.circleAvatarLeftViewLayout(headerFormat)
//
//        expanderWidget.format().avatarText.doMaybe   { avatarText ->
//        expanderWidget.format().avatarFormat.doMaybe { avatarFormat ->
//            layout.addView(this.circleTextAvatarView(avatarText, avatarFormat))
//        } }
//
//        val groupsLayout = this.circleAvatarGroupsViewLayout()
//
//        val groups = expanderWidget.headerGroups(entityId)
//        groups.forEach {
//            //val view = it.view(entityId, context, expanderWidget.groupContext)
//            val view = it.view(entityId, context, groupContext)
//            groupsLayout.addView(view)
//        }
//
//        layout.addView(groupsLayout)
//
//        return layout
//    }
//
//
//    private fun circleAvatarLeftViewLayout(headerFormat : TextFormat) : LinearLayout
//    {
//        val layout                  = LinearLayoutBuilder()
//
//        layout.width                = LinearLayout.LayoutParams.MATCH_PARENT
//        layout.height               = LinearLayout.LayoutParams.WRAP_CONTENT
//
//        layout.orientation          = LinearLayout.HORIZONTAL
//
//        layout.paddingSpacing       = headerFormat.elementFormat().padding()
//        layout.marginSpacing        = headerFormat.elementFormat().margins()
//
//        layout.gravity              = Gravity.CENTER_VERTICAL
//
//        return layout.linearLayout(context)
//    }
//
//
//    private fun circleTextAvatarView(text : String, format : TextFormat) : LinearLayout
//    {
//        // (1) Declarations
//        // -------------------------------------------------------------------------------------
//
//        val layout                  = LinearLayoutBuilder()
//        val textView                = TextViewBuilder()
//
//        // (2) Layout
//        // -------------------------------------------------------------------------------------
//
//        layout.widthDp              = 48
//        layout.heightDp             = 48
//
//        layout.gravity              = Gravity.CENTER
//
//        layout.corners              = Corners(24.0, 24.0, 24.0, 24.0)
//
//        layout.backgroundColor      = colorOrBlack(format.elementFormat().backgroundColorTheme(), entityId)
//
//        layout.child(textView)
//
//        // (3) Text
//        // -------------------------------------------------------------------------------------
//
//        textView.width              = LinearLayout.LayoutParams.WRAP_CONTENT
//        textView.height             = LinearLayout.LayoutParams.WRAP_CONTENT
//
//        textView.text               = text
//
//        format.styleTextViewBuilder(textView, entityId, context)
//
//        return layout.linearLayout(context)
//    }
//
//
//    private fun circleAvatarGroupsViewLayout() : LinearLayout
//    {
//        val layout                  = LinearLayoutBuilder()
//
//        layout.width                = LinearLayout.LayoutParams.MATCH_PARENT
//        layout.height               = LinearLayout.LayoutParams.WRAP_CONTENT
//
//        return layout.linearLayout(context)
//    }
//
//
//    // HEADER > Checkbox Left
//    // -----------------------------------------------------------------------------------------
//
//    private fun checkboxLeftHeaderView() : LinearLayout
//    {
//        val layout = this.headerViewLayout()
//
//        val checkboxLayout = this.checkboxLayout()
//        val checkboxLayoutId = Util.generateViewId()
//        checkboxLayout.id = checkboxLayoutId
//        expanderWidget.checkboxLayoutId = checkboxLayoutId
//
//        checkboxLayout.addView(this.checkboxView())
//
//        layout.addView(checkboxLayout)
//
//        val headerLayout = checkboxLeftHeaderGroupsLayout()
//
//        val groups = expanderWidget.headerGroups(entityId)
//        groups.forEach {
//            val view = it.view(entityId, context, groupContext)
//            headerLayout.addView(view)
//        }
//
//        layout.addView(headerLayout)
//
//        return layout
//    }
//
//
//    private fun checkboxLeftHeaderGroupsLayout() : LinearLayout
//    {
//        val layout                  = LinearLayoutBuilder()
//
//        layout.width                = LinearLayout.LayoutParams.MATCH_PARENT
//        layout.height               = LinearLayout.LayoutParams.WRAP_CONTENT
//
//        return layout.linearLayout(context)
//    }
//
//
//    private fun checkboxLayout() : LinearLayout
//    {
//        val layout = LinearLayoutBuilder()
//
//        val headerFormat = if (this.isOpen)
//            expanderWidget.format().headerOpenFormat()
//        else
//            expanderWidget.format().headerClosedFormat()
//
//        val labelFormat = if (this.isOpen)
//            expanderWidget.format().headerLabelOpenFormat()
//        else
//            expanderWidget.format().headerLabelClosedFormat()
//
//        layout.width                = LinearLayout.LayoutParams.WRAP_CONTENT
//        layout.height               = LinearLayout.LayoutParams.WRAP_CONTENT
//
//        layout.padding.topDp        = headerFormat.elementFormat().padding().topDp()
//        layout.padding.bottomDp     = headerFormat.elementFormat().padding().bottomDp()
//        layout.padding.leftDp       = headerFormat.elementFormat().padding().leftDp()
//        layout.padding.rightDp      = 8f
//
//        layout.onClick              = View.OnClickListener {
//            Router.send(MessageSheetUpdate(ExpanderWidgetUpdateToggle(expanderWidget.widgetId())))
//        }
//
//        return layout.linearLayout(context)
//    }
//
//
//    private fun checkboxView() : LinearLayout
//    {
//        // (1) Declarations
//        // -------------------------------------------------------------------------------------
//
//        val layout = LinearLayoutBuilder()
//        val icon   = ImageViewBuilder()
//
//        val headerFormat = if (this.isOpen)
//            expanderWidget.format().headerOpenFormat()
//        else
//            expanderWidget.format().headerClosedFormat()
//
//        val isSelected = expanderWidget.isSelected(entityId)
//
//        // (2) Layout
//        // -------------------------------------------------------------------------------------
//
//        layout.width                = LinearLayout.LayoutParams.WRAP_CONTENT
//        layout.height               = LinearLayout.LayoutParams.WRAP_CONTENT
//
//        if (isSelected) {
//            layout.backgroundResource   = R.drawable.bg_widget_expander_checkbox_checked
//        }
//        else {
//            layout.backgroundResource   = R.drawable.bg_widget_expander_checkbox
//        }
//
//        layout.gravity              = Gravity.CENTER
//
//        layout.child(icon)
//
//        // (3 A) Icon
//        // -------------------------------------------------------------------------------------
//
//        icon.id                     = R.id.dialog_list_editor_checkbox_icon
//
//        icon.widthDp                = 16
//        icon.heightDp               = 16
//
//        icon.image                  = R.drawable.icon_check_bold
//
//        val iconColorTheme = ColorTheme(setOf(
//                ThemeColorId(ThemeId.Dark, ColorId.Theme("light_green_12")),
//                ThemeColorId(ThemeId.Light, ColorId.Theme("white"))))
//        icon.color                  = colorOrBlack(iconColorTheme, entityId)
//
//        if (isSelected)
//        {
//            icon.visibility             = View.VISIBLE
//        }
//        else
//        {
//            icon.visibility             = View.GONE
//        }
//
//        return layout.linearLayout(context)
//    }
//
//
//
//}
